jps/model-impl/src/com/intellij/openapi/fileTypes/impl/FileTypeAssocTable.java

```diff
@@ -120,7 +139,15 @@ void removeAssociation(@NotNull FileNameMatcher matcher, @Nullable T type) {
       }
       return;
     }
-    myMatchingMappings.removeIf(assoc -> matcher.equals(assoc.getFirst()) && (type == null || type.equals(assoc.getSecond())));
+
+    // Since we need to sort on insertion for IJPL-149806, we allow only one thread to modify the mapping list at a time
+    myMatchingMappingsLock.lock();
```

Comment:
please don't sync, that's why COWAL was here

Reply:
See my comment for https://github.com/JetBrains/intellij-community/pull/2792/files#r1677694484. Maybe I'm wrong, but my understanding is that if `addAssociation` and `removeAssociation` can ever get called from different threads they could mess up the matcher list.

Also, there's one more thing I noticed — but probably benign.

The JB's COW arraylist doesn't overwrite `removeIf` to make it atomic, so it uses the `Collection` version that gets the iterator (which as a copy of the array), iterates over it testing the predicate, and if the predicate matches it calls `remove` on the iterator (c.f. https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Collection.java#L576-L587).

Then JB's COW arraylist just runs `remove` for each single element (c.f. https://github.com/JetBrains/intellij-community/blob/0fa7fb3581d4f15365c9365bb68885595bfc9d67/platform/util/src/com/intellij/util/containers/LockFreeCopyOnWriteArrayList.java#L899-L908), so other threads can see the state of the list in the middle of deletion. It's probably not a big deal if we're only removing stuff, but figured I'd mention it anyway, if I started explaining what I've done here and why.

Reply:
yeah, we don't have to worry about multithreading here, addAssociation/removeAssociation don't get called in parallel.